x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java

```diff
@@ -329,23 +362,81 @@ public void getAuthenticationAndMetaData(String token, ActionListener<Tuple<Auth
+                                    onFailure.accept(new IllegalStateException(
+                                        "token document is missing the access_token field"));
+                                } else if (accessTokenSource.containsKey("user_token") == false) {
+                                    onFailure.accept(new IllegalStateException(
+                                        "token document is missing the user_token field"));
+                                } else {
+                                    Map<String, Object> userTokenSource =
+                                        (Map<String, Object>) accessTokenSource.get("user_token");
+                                    listener.onResponse(UserToken.fromSourceMap(userTokenSource));
+                                }
+                            } else {
+                                onFailure.accept(
+                                    new IllegalStateException("token document is missing and must be present"));
+                            }
+                        }, e -> {
+                            // if the index or the shard is not there / available we assume that
+                            // the token is not valid
+                            if (isShardNotAvailableException(e)) {
+                                logger.warn("failed to get token [{}] since index is not available", userTokenId);
+                                listener.onResponse(null);
+                            } else {
+                                logger.error(new ParameterizedMessage("failed to get token [{}]", userTokenId), e);
+                                listener.onFailure(e);
+                            }
+                        }), client::get);
+                });
+        }
+    }
+
+    /*
+     * If needed, for tokens that were created in pre 7.1.0 cluster, it asynchronously decodes the string representation of a {@link
+     * UserToken}. The process for this is asynchronous as we may need to compute a key, which can be computationally expensive
+     * so this should not block the current thread, which is typically a network thread. A second reason for being asynchronous is that
+     * we can restrain the amount of resources consumed by the key computation to a single thread.
      */
     void decodeToken(String token, ActionListener<UserToken> listener) throws IOException {
         // We intentionally do not use try-with resources since we need to keep the stream open if we need to compute a key!
         byte[] bytes = token.getBytes(StandardCharsets.UTF_8);
         StreamInput in = new InputStreamStreamInput(Base64.getDecoder().wrap(new ByteArrayInputStream(bytes)), bytes.length);
-        if (in.available() < MINIMUM_BASE64_BYTES) {
-            logger.debug("invalid token");
-            listener.onResponse(null);
+        // the token exists and the value is at least as long as we'd expect
```

Comment:
I do not think this comment is correct; we haven't checked the length at this point